pkg/runctl/controller_test.go

```diff
@@ -151,6 +125,114 @@ func Test_Controller_syncHandler_givesUp_onPipelineRunNotFound(t *testing.T) {
 	assert.NilError(t, err)
 }
 
+func newController(runs ...*api.PipelineRun) (*Controller, *fake.ClientFactory) {
+	cf := fake.NewClientFactory(fake.ClusterRole(string(runClusterRoleName)))
+	cs := cf.StewardClientset()
+	cs.PrependReactor("create", "*", fake.NewCreationTimestampReactor())
+	client := cf.StewardV1alpha1()
+	for _, run := range runs {
+		client.PipelineRuns(run.GetNamespace()).Create(run)
+	}
+	metrics := metrics.NewMetrics()
+	controller := NewController(cf, metrics)
+	controller.pipelineRunFetcher = k8s.NewClientBasedPipelineRunFetcher(client)
+	return controller, cf
+}
+
+func getAPIPipelineRun(cf *fake.ClientFactory, name, namespace string) (*api.PipelineRun, error) {
+	cs := cf.StewardClientset()
+	return cs.StewardV1alpha1().PipelineRuns(namespace).Get(name, metav1.GetOptions{})
+}
+
+func Test_Controller_syncHandler(t *testing.T) {
+	for _, test := range []struct {
+		name            string
+		pipelineSpec    api.PipelineSpec
+		currentStatus   api.PipelineStatus
+		runManagerError error
+		expectedResult  api.Result
+		expectedState   api.State
+		expectedMessage string
+	}{
+		{name: "new_ok",
+			pipelineSpec:    api.PipelineSpec{},
+			currentStatus:   api.PipelineStatus{},
+			runManagerError: nil,
+			expectedResult:  api.ResultUndefined,
+			expectedState:   api.StateWaiting,
+		},
+		{name: "new_fail",
+			pipelineSpec:    api.PipelineSpec{},
+			currentStatus:   api.PipelineStatus{},
+			runManagerError: fmt.Errorf("expected"),
+			expectedResult:  api.ResultErrorInfra,
+			expectedState:   api.StateCleaning,
+			expectedMessage: ".*Failed to create run namespace.*",
+		},
+		{name: "new_missing_secret",
+			pipelineSpec: api.PipelineSpec{
+				Secrets: []string{"secret1"},
+			},
+			currentStatus:   api.PipelineStatus{},
+			runManagerError: nil,
+			expectedResult:  api.ResultErrorContent,
+			expectedState:   api.StateCleaning,
+			expectedMessage: "failed to copy secrets: .*",
+		},
+		{name: "waiting_fail",
+			pipelineSpec: api.PipelineSpec{},
+			currentStatus: api.PipelineStatus{
+				State: api.StateWaiting},
+			runManagerError: fmt.Errorf("expected"),
+			expectedResult:  api.ResultErrorInfra,
+			expectedState:   api.StateCleaning,
+		},
+		{name: "skip_new",
+			pipelineSpec: api.PipelineSpec{},
+			currentStatus: api.PipelineStatus{
+				State: api.StateNew},
+			runManagerError: nil,
+			expectedResult:  "",
+			expectedState:   api.StateNew,
+		},
+		{name: "skip_finished",
+			pipelineSpec: api.PipelineSpec{},
+			currentStatus: api.PipelineStatus{
+				State: api.StateFinished},
+			runManagerError: nil,
+			expectedResult:  "",
+			expectedState:   api.StateFinished,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			test := test
```

Comment:
?

Reply:
Must be moved before `t.Run(...)`.

The function passed to `t.Run()` is a closure that captures the variables from the scope where it was defined. If it it would just use the loop variable `test`, it would see its current value at the time the function executes, which is the value from the last iteration. That means each subtest would test the last testcase.

By defining a _new_ variable `test` within the loop body, initialized with the current value of the loop variable `test` (these are two different variables in different scopes), each test function will capture this variable instead of the loop variable and therefore run the respective testcase as expected.

Reply:
I think it is sufficient to do this before the t.Parallel() is called